sgtm/chaos/write_event_data_to_bigquery.tpl

```diff
@@ -0,0 +1,502 @@
+              }
+            ],
+            "isUnique": true,
+            "valueValidators": []
+          }
+        ],
+        "valueValidators": [
+          {
+            "type": "NON_EMPTY",
+            "errorMessage": "Please select at least 1 parameter"
+          }
+        ]
+      }
+    ]
+  }
+]
+
+
+___SANDBOXED_JS_FOR_SERVER___
+
+/**
+ * @license
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+const addEventCallback = require('addEventCallback');
+const BigQuery = require('BigQuery');
+const getEventData = require('getEventData');
+const getTimestampMillis = require('getTimestampMillis');
+const JSON = require('JSON');
+const log = require('logToConsole');
+const Math = require('Math');
+
+
+const currentTimestamp = getTimestampMillis();
+const event_name = getEventData('event_name') || null;
+const ga_session_id = getEventData('ga_session_id') || null;
+const ga_session_number = getEventData('ga_session_number') || null;
+const client_id = getEventData('client_id') || null;
+const user_id = getEventData('user_id') || null;
+const user_agent = getEventData('user_agent') || null;
+const ip_address = getEventData('ip_override') || null;
+const event_location = getEventData('event_location') || null;
+const event_location_country = event_location ? event_location.country : null;
+const event_location_region = event_location ? event_location.region : null;
+const page_location = getEventData('page_location') || null;
+const page_title = getEventData('page_title') || null;
+const page_referrer = getEventData('page_referrer') || null;
+const language = getEventData('language') || null;
+const screen_resolution = getEventData('screen_resolution') || null;
+const value = getEventData('value') || 0;
+const currency = getEventData('currency') || null;
+const transaction_id = getEventData('transaction_id') || null;
+const x_ga_dma = getEventData('x-ga-dma') || null;
+const x_ga_dma_cps = getEventData('x-ga-dma_cps') || null;
+const x_ga_gcd = getEventData('x-ga-gcd') || null;
+const x_ga_gcs = getEventData('x-ga-gcs') || null;
+const x_ga_npa = getEventData('x-ga-npa') || null;
+const x_ga_measurement_id = getEventData('x-ga-measurement_id') || null;
+const x_ga_page_id = getEventData('x-ga-page_id') || null;
+const ga_debug_mode = getEventData('debug_mode') || null;
+
+
+// function for creating event_date
+function createDateString(currentTimestamp) {
```

Comment:
This has a lot of complex logic with a lot of hard coded constants that aren't explained. Makes it hard for an engineer to follow. I think you should write some tests for this if you don't mind. I can help you if needed.

Reply:
Will take you up on that

Reply:
Ok I've had a crack at this and got it working with these changes:

Test Setup:
```
const logToConsole = require('logToConsole');

// The tag configuration for the tests
let mockData = {
  'gcpProjectId': 'my-project',
  'datasetId': 'my_dataset',
  'tableId': 'MyTable',
  'attributesList': [
    {'attribute': 'event_timestamp'},
    {'attribute': 'date'},
  ]
};

/**
 * Add mock BigQuery library to the test which tests date is expected value.
 * @param {string} expectedDate - the expected value for the date.
 */
function addMockBigQueryWithDateCheck(expectedDate){
  mock('BigQuery', () => {
    return {
      'insert': (connectionInfo, rows) => {
        logToConsole(connectionInfo);
        logToConsole(rows);
        assertThat(rows).hasLength(1);
        assertThat(rows[0].date).isEqualTo(expectedDate);
      }
    };
  });
}

/**
 * Mock the timestamp functionality and return the given timestamp in millis.
 * @param {number} timestampMillis - the timestamp in millis to use
 */
function addMockGetTimestampMillis(timestampMillis){
  mock('getTimestampMillis', () => {
    return timestampMillis;
  });
}
```

Test createDateString formats dates as expected: 
```
addMockGetTimestampMillis(1704099600000);
addMockBigQueryWithDateCheck('20240101');

runCode(mockData);

addMockGetTimestampMillis(1709164800000);
addMockBigQueryWithDateCheck('20240229');

runCode(mockData);
```

However, you also need to change the code:
```
...
// pushing into object value for each parameter user selected to write in BigQuery
let outputObject = {};
attributesNames.forEach(param => {
  if(eventData.hasOwnProperty(param)) {
    outputObject[param] = eventData[param];
  } 
});

// Mock API returns a function, whereas usual import is object.
// This logic enables support for tests within the function. See
// https://developers.google.com/tag-platform/tag-manager/server-side/api#mock
let bigquery = BigQuery;
if (getType(BigQuery) === "function"){
  bigquery = BigQuery();
}


//addEventCallback(() => {
  bigquery.insert({
    projectId: data.gcpProjectId,
    datasetId: data.datasetId,
    tableId: data.tableId,
  }, [outputObject], {}, () => {
    log('BigQuery Success: ', [outputObject]);
  }, (errors) => {
    log('BigQuery Failure: ', JSON.stringify(errors));
  });
//});

```
Note that I've commented out the `addEventCallback()` function. I'm not clear if you need that. If you don't, please remove. If you do need it we'll need to figure out how to rework the test to work with this.